Feature/deconvolution

al-de opened the pull request:
### Changes proposed in this pull request:
- implementation of deconvolve function in the dsp module
- basic testing for this deconvolve function

### Labels
- **feature**
- **dsp**

f-brinkmann reviewed (changes requested):
Thanks for the effort - this looks mostly fine to me. I added `deconv` to `dsp/__init__.py` to make it appear in the documentation. Now the docstring throws warning when you run `make html` inside the `docs` folder. But they are probably easy to fix.

One question for discussion: Should we have a parameter to cut potential distortion from the end of the IR?

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+    .. math::
+    \\H(\\omega) = Y(\\omega)\\hat{X(\\omega)}.
+
+    For more information, refer to [1]_
```
This makes sense, thanks for pointing me to the relevant section in CONTRIBUTING.rst :)

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+
+    References
+    -----------
+    .. [1] S. Mueller and P. Masserani "Transfer function measurement with
```
check :)

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+    where :math:`X(\\omega)` is the excitation signal and :math:`Y(\\omega)`
+    the measured signal. Regulated inversion is used to avoid numerical issues
+    in calculating :math:`\\hat{X(\\omega)} = 1/X(\\omega)` for small values of
+    :math:`X(\\omega)` (see :py:func:`pyfar.dsp.regulated_spectrum_inversion`).
```
Thanks for this tip, as I'm still getting warm with the docstrings, I am glad about every new hint. :)

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+        The excitation signal, used to perform the measurement.
+        The excitation signal is zero padded, if it is shorter than the
+        measurement signal.
+    kwargs : Parameters for regu_inversion, can include:
+        freq_range, regu_outside, regu_inside, regu_final
+        refer to function regularized_spectrum_inversion for deatiled
+        documentation.
```
check

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+
+    Parameters
+    ----------
+    measurement : pyfar.Signal
```
check

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+    if not isinstance(measurement, pyfar.Signal):
+        raise TypeError('Input data has to be of type: Signal.')
+    if not isinstance(excitation, pyfar.Signal):
+        raise TypeError('Input data has to be of type: Signal.')
```
I implemented the more explicit version with my last commit.

al-de commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+
+    # Check if both signals have the same length,
+    # if not: bring them to the same length
+    if measurement.n_samples > excitation.n_samples:
```
That's true, this was an artifact from the time before the first release of pyfar, as there was no pad_zeros function.
I fixed this with the following change in my last commit for the case `measurement.n_samples > excitation.n_samples` and vice versa for the case `measurement.n_samples < excitation.n_samples`:

`excitation = pyfar.dsp.pad_zeros(excitation, (measurement.n_samples -
                                                      excitation.n_samples))`

al-de commented on tests/test_dsp_deconv.py:
```diff
@@ -0,0 +1,52 @@
+
+def test_input_type_error():
+    """Test assertions by passing non Signal-Type"""
+    with pytest.raises(TypeError):
```
Thanks for this input, I changed this with my last commit and used the match parameter to check the exact error message.

al-de commented:
Thanks a lot for your very detailed feedback, it is a great help.

I resolved all requested changes by committing your suggestions respectively writing the code you described.
Next step I will look into these warnings thrown when running `make html` inside `docs`.
I left all comments unresolved that need some further discussion.

sikersten commented on pyfar/dsp/dsp.py:
```diff
@@ -1396,3 +1396,100 @@ def time_shift(signal, shift, unit='samples'):
+
+    # multiply measurement signal with regularized inversed excitation signal
+    # to get the transfer function
+    result = measurement * regularized_spectrum_inversion(excitation, **kwargs)
```
Theoretically, also signals with other fft_norms than `'none'` could be deconvolved, or did I miss something? If so, I'd suggest to remove the enforcement at all and leave it up to the user to be responsible for the correct `fft_norm` of the input. Let's discuss in the weekly!

sikersten reviewed (commented):
Functionality looks good to me, thanks for implementing! :)
One thing we can discuss: maybe a renaming of the parameters `measurement` and `excitation` to something like `output` and `input` would make sense to indicate that this function does not necessarily needs to be applied to real measured signals, but implements a general approach to obtain a transfer function.
Regarding @f-brinkmann's question to cut distortion: I'd suggest to put that functionality in a different function (isn't it simply `time_window`?). Combining these methods is in my opinion part of a measurement setup (in another package?).

al-de commented:
Thanks a lot for your Feedback. :)

In the latest commits I renamed `measurement`, `excitation` and `result` to `system_output`, `system_input` and `system_response` to fit the naming to a more general system approach.

Regarding the cut distortion topic, I implemented the new parameter `fft_length` for manually setting the `n_samples` both signals are padded to. By default `fft_length` is set to `None`, which results in the same behavior as before, where only the shorter signal is padded to the length of the longer signal.

With the implementation of the correct handling of the `fft_norm` of both signals I will wait until #223 and #224 are closed.

sikersten commented on pyfar/dsp/dsp.py:
```diff
@@ -1395,6 +1395,119 @@ def time_shift(signal, shift, unit='samples'):
+        system_response.comment += f" system output: {system_input.comment}."
+
+    # return the impulse resonse
+    system_response.fft_norm = pyfar.classes.audio._match_fft_norm(
+        system_output.fft_norm, system_input.fft_norm)
```
Deconvolution means division in frequency domain, therefore:
```suggestion
    system_response.fft_norm = pyfar.classes.audio._match_fft_norm(
        system_output.fft_norm, system_input.fft_norm, division=True)
```
Adding a small test to check for the correct call might make sense here :)

al-de replied:
You are totally right, I thought that the `division=True` will be handled in the `regularized_spectrum_inversion`, but as the resulting fft_norm is determined directly from the parameters of this function, your suggestion is the right way of handling the fft_norm in this case.
Thanks for pointing this out. :)

al-de commented:
Thanks a lot for merging develop into this branch to solve the conflicts. :)

It seems like now Sphinx throws an error that is related to the convolution function:
`lib/python3.8/site-packages/matplotlib/sphinxext/plot_directive.py", line 484, in run_code
    exec(code, ns)
  File "<string>", line 4, in <module>
  File "/home/alex/Git/pyfar-pyfar/pyfar/pyfar/dsp/dsp.py", line 1592, in convolve
    raise ValueError("The sampling rates do not match")
ValueError: The sampling rates do not match`

If I get it right, there are two Signals with different Sampling rates (1000 & 0.1 Hz) generated for the example and than convolved. This lead to the ValueError.

`>>> s1 = pf.Signal([1, 0.5, 0.5], 1000)`
`>>> s2 = pf.Signal([1,-1], .1)`

Should we solve this with a separate pull request on develop or include this small fix in this pull here?

sikersten commented on pyfar/dsp/dsp.py:
```diff
@@ -1395,6 +1395,119 @@ def time_shift(signal, shift, unit='samples'):
+        raise ValueError("The fft_length can not be shorter than" +
+                         "system_input.n_samples.")
+
+    # Check if both signals have the same length as ftt_length,
+    # if not: bring them to the same length by padding with zeros
+    if system_output.n_samples < fft_length:
+        # Add Zeros to system_input
+        system_output = pyfar.dsp.pad_zeros(system_output,
+                                            (fft_length -
+                                             system_output.n_samples))
+
+    if system_input.n_samples < fft_length:
+        # Add Zeros to system_output
+        system_input = pyfar.dsp.pad_zeros(system_input,
+                                           (fft_length -
+                                            system_input.n_samples))
```
As it is checked above that fft_length>=n_samples, this could be simplified to:
```suggestion
    system_output = pyfar.dsp.pad_zeros(
        system_output, (fft_length - system_output.n_samples))
    system_input = pyfar.dsp.pad_zeros(
        system_input, (fft_length - system_input.n_samples))
```

al-de replied:
check, good suggestion. :)

sikersten commented on tests/test_dsp_deconv.py:
```diff
@@ -0,0 +1,75 @@
+
+from pyfar.dsp.dsp import deconvolve
+
+
```
I'd suggest to add a `test_default`, which checks the correct results (incl. fft_norm etc.) for calling `deconvolve` with default parameters.

al-de replied:
good idea, I will add a `test_default`.

sikersten commented:
@al-de thanks for noting the sphinx error due to convolution example, I'll fix that on the develop branch

al-de commented:
Thanks for your feedback and the fix on develop.
I merged the fix to this branch and implemented your feedback.
